README.md
# React Azure MSAL Security

### Azure Active Directory security module using MSAL

React library for application security (using MSAL) that eases application protection with Azure session JWT.
Exposes multiples hooks for login, logout and secure components.

## Structure 📋

```bash
├── README.md
├── LICENCE.md
├── CHANGELOG.md
├── .vscode/ # vscode shared development config
├── src/
│   ├── security/
│   │   ├── config/
│   │   │   ├── aad.config.js # MSAL config initialization
│   │   │   └── aad.types.js # AAD and Ms Graph constants
│   │   ├── services/
│   │   │   ├── aad-graph.service.js # basic graph service for user info and avatar
│   │   │   └── aad.service.js # service for handle login, sso and token acquisition
│   │   ├── auth.hook.js # module hooks
│   │   ├── cache.util.js # util for persist graph info
│   │   └── observer.util.js # observer pattern handler
│   └── index.js
├── package.json
├── jsconfig.js
├── .babelrc
├── .eslintrc.json
└── .prettierrc.json
```

| Files                  | Description                                                                            |
| ---------------------- | -------------------------------------------------------------------------------------- |
| `aad.config.js`        | contains context config                                                                |
| `aad.types.js`         | contains MSAL anf Microsoft Graph constants                                            |
| `aad.service.js`       | main service. Handles MSAL context, session state, login, logout and token acquisition |
| `aad-graph.service.js` | handles Microsoft Graph calls, like user detailed info and profile avatar              |
| `auth.hooks.js`        | exposed hooks for login, logout and secure components                                  |
| `index.js`             | exports router, hooks and routes handler/service                                       |

## Features 🎉

✅ Centralized and synchronized session state handling for any authentication event.

✅ Automatic and manual login hooks.

✅ Logout, state and token acquisition hooks.

✅ Refresh token route handler.

✅ User specific profile info and avatar hooks, using Microsoft Graph.

⚠️ Easy to use! but a little hard to customize, i.e. user graph info.

## How To Configure 📖

First, you should [register](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) a new Azure Active Directory Application in Azure Portal.

In **Authentication** section add a new platform, selecting Single-Page Application (SPA) and setting up Redirect URIs only, with host (base URL) and token auth URL, for example, for dev:

-   https://localhost:3000/
-   https://localhost:3000/auth

Also, in this section, enable **Access tokens** and **ID tokens** in **Implicit grant** sub-section.

Finally, you should configure permissions in **API permissions** section, with the minimum permission _User.Read_, although ideally _profile_ and _openid_ for correct token acquisition.

## How To Use 💡

This module, exposes many hooks for handle login, logout and secure React components.

### App configuration

In your **App.jsx** or **App.router.jsx** you should initialize the authentication service.

```javascript
import { AuthenticationService } from '@calvear/react-azure-msal-security';

// MSAL config.
const authConfig = {
    clientId: '2a85c521-02fc-4796-8ecc-eaa13eee2e7b', // registered app id from azure
    tenantId: 'ba3947ca-abb7-402e-b1d1-c9284608f497', // maybe common, organizations or consumers also
    loginActionRedirect: '/',
    logoutActionRedirect: null,
    tokenRefreshUri: '/auth', // should exists a blank route in your app
    tokenRenewalOffset: 120,
    navigateToRequestAfterLogin: true,
};

// initializes Microsoft Active Directory authentication service.
AuthenticationService.init(authConfig);

// main react component
export default () => {
    return (
        <div>
            <h1>Welcome to My App</h1>
        </div>
    );
};
```

| Parameters                             | Description                                                     |
| -------------------------------------- | --------------------------------------------------------------- |
| `[config]`                             | settings                                                        |
| `[config.tenantId]`                    | organization Azure client id                                    |
| `[config.clientId]`                    | application Azure client id                                     |
| `[config.loginActionRedirect]`         | (default: '/') - redirect path after login                      |
| `[config.logoutActionRedirect]`        | (default: null) - redirect path after logout                    |
| `[config.tokenRefreshUri]`             | (default: '/auth') - path for renew auth token                  |
| `[config.tokenRenewalOffset]`          | (default: 120) token renewal interval                           |
| `[config.navigateToRequestAfterLogin]` | (default: true) - if app redirects to previous path after login |
| `[config.infoCacheDurationInDays]`     | (default: 1) - days for store user info cached                  |
| `[config.photoCacheDurationInDays]`    | (default: 3) - days for store user photo cached                 |
| `[disabled]`                           | (default: false) - if authentication is disabled globally       |

For tenantId also see [MSAL Client Config](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-client-application-configuration)

### Token acquisition and blank page/route

For a correct authentication and token renewal operation you should define a blank page or route in your application.
This route will be used as iframe for retrieve or renew the token on authentication.

For example, if you're using [react-spa-routerizer](https://github.com/calvear93/react-spa-routerizer)

```javascript
// routes.js

export default [
    ...,
    // blank html page for load authentication iframe to refresh the token,
    // also, you should set tokenRefreshUri as '/auth' route.
    {
        key: 'auth-page',
        title: 'Authenticating',
        path: '/auth',
        Child: () => null
    },
    ...
];
```

### ☑️ Check Authentication State

```javascript
import { useAuthenticationState } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const { authenticated, authenticating, error } = useAuthenticationState();

    return (
        <div>
            <h1>User is authenticated?: {authenticated ? 'yeah' : 'no'}</h1>
        </div>
    );
};
```

| Returning Modules      | Description                      |
| ---------------------- | -------------------------------- |
| `state`                | object with authentication state |
| `state.authenticated`  | if user is authenticated         |
| `state.authenticating` | if service is authenticating     |
| `state.error`          | error object                     |

### ☑️ Automatic Login

```javascript
import { useAuthentication } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const { authenticated, authenticating, error } = useAuthentication();

    if (authenticating) return <div>Authenticating...</div>;

    if (!authenticated) return <div>403: Not Authorized - {error.message}</div>;

    return (
        <div>
            <h1>Welcome to My App</h1>
        </div>
    );
};
```

| Returning Modules      | Description                      |
| ---------------------- | -------------------------------- |
| `state`                | object with authentication state |
| `state.authenticated`  | if user is authenticated         |
| `state.authenticating` | if service is authenticating     |
| `state.error`          | error object                     |

| Parameters           | Description                                            |
| -------------------- | ------------------------------------------------------ |
| `[config]`           | settings                                               |
| `[config.loginType]` | (default: loginRedirect) - loginRedirect or loginPopup |

### ☑️ Manual Login

```javascript
import { Redirect } from 'react-router-dom';
import { useAuthenticationState, useLogin } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const login = useLogin();
    const { authenticated, authenticating, error } = useAuthenticationState();

    if(authenticating)
        return <div>Authenticating...</div>;

    if(!authenticated)
        return <div>403: Not Authorized - {error.message}</div>;

    if(authenticated)
        return <Redirect to='/home'>;

    return (
        <div>
            <button onClick={login}>Log In</button>
        </div>
    );
}
```

| Returning Modules | Description                                       |
| ----------------- | ------------------------------------------------- |
| `login`           | (login) function for trigger login/authentication |

| Parameters    | Description                                          |
| ------------- | ---------------------------------------------------- |
| `[loginType]` | (default: loginRedirect) loginRedirect or loginPopup |

### ☑️ Logout

```javascript
import { useLogout } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const logout = useLogout();

    return (
        <div>
            <button onClick={logout}>Log Out</button>
        </div>
    );
};
```

| Returning Modules | Description         |
| ----------------- | ------------------- |
| `logout`          | function for logout |

### ☑️ Acquire Token

You can acquire a JWT access token for API securing.

```javascript
import { useState, useEffect } from 'react';
import { useAcquireToken } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const [token, setToken] = useState();
    const { authenticated } = useAuthentication();
    // use it only if authentication was succeeded
    const acquireToken = useAcquireToken();

    useEffect(() => {
        if (authenticated) {
            acquireToken().then((tkn) => setToken(tkn));
        }
    }, [authenticated]);

    return (
        <div>
            <h1>Welcome to My App</h1>
        </div>
    );
};
```

| Returning Modules | Description                                |
| ----------------- | ------------------------------------------ |
| `acquireToken`    | async function for acquire an access token |

| Parameters            | Description                                                     |
| --------------------- | --------------------------------------------------------------- |
| `[forceTokenRefresh]` | (default: false) - forces to renew token from active directory. |

### ☑️ Graph Info

You can retrieves user account detailed info and profile avatar from Microsoft Graph api with hooks.

```javascript
import { useAccountInfo } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const { loading, info, error } = useAccountInfo();

    if (loading) return <div>Loading User Info...</div>;

    if (error) return <div>User info cannot be loaded: {error.message}</div>;

    return (
        <div>
            <h1>User Info</h1>
            <h3>Name: {info.displayName}</h3>
        </div>
    );
};
```

| Returning Modules | Description            |
| ----------------- | ---------------------- |
| `state`           | object with info state |
| `state.loading`   | if info is loading     |
| `state.info`      | account info           |
| `state.error`     | error object           |

```javascript
import { useAccountAvatar } from '@calvear/react-azure-msal-security';

// react component
export default () => {
    const { loading, avatar, error } = useAccountAvatar();

    if (loading) return <div>Loading User Avatar...</div>;

    if (error) return <div>User avatar cannot be loaded: {error.message}</div>;

    return (
        <div>
            <img alt="user-avatar" src={avatar} />
        </div>
    );
};
```

| Returning Modules | Description              |
| ----------------- | ------------------------ |
| `state`           | object with avatar state |
| `state.loading`   | if avatar is loading     |
| `state.avatar`    | account avatar in base64 |
| `state.error`     | error object             |

| Parameters   | Description                                                                                                               |
| ------------ | ------------------------------------------------------------------------------------------------------------------------- |
| `[size]` | (default: 648x648) avatar sizes. Values maybe 48x48, 64x64, 96x96, 120x120, 240x240, 360x360, 432x432, 504x504 or 648x648 |

## Linting 🧿

Project uses ESLint, for code formatting and code styling normalizing.

-   **eslint**: JavaScript and React linter with Airbnb React base config and some other additions.
-   **prettier**: optional Prettier config.

For correct interpretation of linters, is recommended to use [Visual Studio Code](https://code.visualstudio.com/) as IDE and install the plugins in .vscode folder at 'extensions.json', as well as use the config provided in 'settings.json'

## Changelog 📄

For last changes see [CHANGELOG.md](CHANGELOG.md) file for details.

## Built with 🛠️

-   [React](https://reactjs.org/) - the most fabulous JavaScript framework.
-   [MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-js) - Microsoft Authentication Library for JavaScript.
-   [axios](https://github.com/axios/axios) - Promise based HTTP client.

## License 📄

This project is licensed under the MIT License - see [LICENSE.md](LICENSE.md) file for details.

---

⌨ by [Alvear Candia, Cristopher Alejandro](https://github.com/calvear93)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [2.1.2] - 2021-05-22
-   dependencies updated

## [2.1.1] - 2021-05-19
-   possible acquisition token fix

## [2.0.5] - 2021-04-25
-   improved token acquisition.
-   dependencies updated.

## [2.0.5] - 2021-04-02
-   MSAL updated.

## [2.0.4] - 2021-03-27
-   dependencies updated.

## [2.0.3] - 2021-02-27
-   enabled sourcemaps for compiled code.
-   .eslintignore added for build/dist folders.

## [2.0.2] - 2021-02-18
-   Changed minimum React version to 16.13.1.

## [2.0.1] - 2021-02-10
-   fixed blob handling for avatar in some scenarios.

## [2.0.0] - 2021-02-09
-   state handle reworked. Now, service has a centralized observer handling state changes.
-   added useAuthenticationState that handles in real-time session changes.
-   removed conditional authentication hooks. You can use useEffect for postLogin validations.
-   simplified init function.

## [1.0.4] - 2021-02-09
-   useLogin authenticated from state fixed.

## [1.0.3] - 2021-02-08
-   added useIsAuthenticated hook for simple check authentication state.

## [1.0.2] - 2021-02-08
-   added forceTokenRefresh param in useAcquireToken hook (forces to renew token from active directory).

## [1.0.1] - 2021-02-07
-   readme docs updated.

## [1.0.0] - 2021-02-06
-   MSAL authentication service.
-   Microsoft Graph service.
-   module initialization function from service.
-   hooks for login, logout and conditional login.


src/security/config/aad.cfg.js
/**
 * MSAL Microsoft Authentication configuration file.
 *
 * @see https://azuread.github.io/microsoft-authentication-library-for-js/docs/msal/modules/_configuration_.html
 *
 * Azure Active Directory App Registration should
 * has 'user.read' and 'openid' (for access tokens)
 * scopes allowed.
 *
 * @summary MSAL config file.
 * @author Alvear Candia, Cristopher Alejandro <calvear93@gmail.com>
 *
 * Created at     : 2020-04-15 19:51:39
 * Last modified  : 2020-12-04 15:19:24
 */

import types from './aad.types';

/**
 * Creates a config object for
 * MSAL authentication context.
 *
 * @export
 * @param {object} config MSAL auth config.
 * @param {string} config.tenantId organization Azure Object Id.
 * @param {string} config.clientId application Azure Object Id.
 * @param {string} [config.loginActionRedirect] redirect path after login.
 *  If navigateToRequestAfterLogin is false.
 * @param {string} [config.logoutActionRedirect] redirect path after logout.
 * @param {string} [config.tokenRefreshUri] path for renew auth token.
 *  Should be a empty page (null React component) and should be added to
 *  authentication routes in Azure App Registration.
 * @param {number} [config.tokenRenewalOffset] token renewal interval.
 * @param {string} [config.navigateToRequestAfterLogin] whether app redirects to previous path after login.
 * @param {number} [config.infoCacheDurationInDays] days for store user info cached.
 * @param {number} [config.photoCacheDurationInDays] days for store user photo cached.
 *
 * @returns {object} MSAL config object.
 */
export default function createConfig({
    clientId,
    tenantId,
    loginActionRedirect = '/',
    logoutActionRedirect = null,
    tokenRefreshUri = '/auth',
    tokenRenewalOffset = 120,
    navigateToRequestAfterLogin = true,
    infoCacheDurationInDays = 1,
    photoCacheDurationInDays = 3
})
{
    // login redirect URL.
    const LOGIN_ACTION_REDIRECT = loginActionRedirect && loginActionRedirect !== 'null'
        ? `${window.location.origin}${loginActionRedirect}`
        : window.location.origin;

    // logout redirect URL.
    const LOGOUT_ACTION_REDIRECT = logoutActionRedirect && logoutActionRedirect !== 'null'
        ? `${window.location.origin}${logoutActionRedirect}`
        : window.location.origin;

    // token acquisition route path.
    const TOKEN_REFRESH_URI = tokenRefreshUri && tokenRefreshUri !== 'null'
        ? `${window.location.origin}${tokenRefreshUri}`
        : window.location.origin;

    // offset needed to renew the token before expiry.
    const TOKEN_REFRESH_PERIOD = +tokenRenewalOffset;

    // navigates to request URL after authentication/login instead of redirect URL.
    const NAVIGATE_TO_REQUEST_URL_AFTER_LOGIN = navigateToRequestAfterLogin === true || navigateToRequestAfterLogin === 'true';

    /**
     *  - clientId: Client ID of your app registered with our Application registration portal (https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredAppsPreview) in Microsoft Identity Platform
     *  - authority: You can configure a specific authority, defaults to " " or "https://login.microsoftonline.com/common"
     *  - validateAuthority: Used to turn authority validation on/off. When set to true (default), MSAL will compare the application's authority against well-known URLs templates representing well-formed authorities. It is useful when the authority is obtained at run time to prevent MSAL from displaying authentication prompts from malicious pages.
     *  - knownAuthorities: If validateAuthority is set to True, this will be used to set the Trusted Host list. Defaults to empty array
     *  - redirectUri: The redirect URI of the application, this should be same as the value in the application registration portal.Defaults to `window.location.href`.
     *  - postLogoutRedirectUri: Used to redirect the user to this location after logout. Defaults to `window.location.href`.
     *  - navigateToLoginRequestUrl: Used to turn off default navigation to start page after login. Default is true. This is used only for redirect flows.
     *  - tokenRefreshUri: [CUSTOM] allows to exports token acquisition iframe route for token refresh logic.
     */
    const auth = {
        clientId,
        // https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-client-application-configuration
        authority: `https://login.microsoftonline.com/${tenantId}`,
        validateAuthority: false,
        redirectUri: LOGIN_ACTION_REDIRECT,
        postLogoutRedirectUri: LOGOUT_ACTION_REDIRECT,
        navigateToLoginRequestUrl: NAVIGATE_TO_REQUEST_URL_AFTER_LOGIN,
        tokenRefreshUri: TOKEN_REFRESH_URI
    };

    /**
     * Use this to configure the below cache configuration options.
     *
     * - cacheLocation: Used to specify the cacheLocation user wants to set. Valid values are "localStorage" and "sessionStorage"
     * - storeAuthStateInCookie: If set, MSAL store's the auth request state required for validation of the auth flows in the browser cookies. By default this flag is set to false.
     * - infoCacheDurationInDays: [CUSTOM] days for store user info in cache.
     * - photoCacheDurationInDays: [CUSTOM] days for store user photo in cache.
     */
    const cache = {
        cacheLocation: types.CACHE.LOCAL_STORAGE,
        storeAuthStateInCookie: false,
        infoCacheDurationInDays: +infoCacheDurationInDays,
        photoCacheDurationInDays: +photoCacheDurationInDays
    };

    /**
     * Library specific options.
     *
     * - logger: Used to initialize the Logger object;
     * - loadFrameTimeout: maximum time the library should wait for a frame to load
     * - tokenRenewalOffsetSeconds: sets the window of offset needed to renew the token before expiry
     * - navigateFrameWait: sets the wait time for hidden iFrame navigation
     */
    const system = {
        loadFrameTimeout: 12000,
        tokenRenewalOffsetSeconds: TOKEN_REFRESH_PERIOD,
        navigateFrameWait: 200
    };

    /**
     * App/Framework specific environment support.
     *
     * - unprotectedResources: Array of URI's which are unprotected resources. MSAL will not attach a token to outgoing requests that have these URI. Defaults to 'null'.
     * - protectedResourceMap: This is mapping of resources to scopes used by MSAL for automatically attaching access tokens in web API calls.A single access token is obtained for the resource. So you can map a specific resource path as follows: {"https://graph.microsoft.com/v1.0/me", ["user.read"]}, or the app URL of the resource as: {"https://graph.microsoft.com/", ["user.read", "mail.send"]}. This is required for CORS calls.
     */
    const framework = {
        unprotectedResources: [],
        protectedResourceMap: new Map()
    };

    // MSAL configuration.
    return {
        auth,
        cache,
        system,
        framework
    };
}
